odoo/tools/misc.py

```diff
@@ -1641,3 +1641,21 @@ def hmac(env, scope, message, hash_function=hashlib.sha256):
         message.encode(),
         hash_function,
     ).hexdigest()
+
+
+bots = ["bot", "crawl", "slurp", "spider", "curl", "wget", "facebookexternalhit", "whatsapp", "trendsmapresolver", "pinterest", "instagram"]
+def is_a_bot(user_agent):
+    """
+    Try to match if the given user agent is a bot or not based on a list of
+    bot identifiers.
+
+        :param str user_agent: user agent string to check
+        :return: True if the user agent is a bot, else False
+        """
+    user_agent = user_agent.lower()
+    # We don't use regexp and ustr voluntarily
+    # timeit has been done to check the optimum method
+    try:
+        return any(bot in user_agent for bot in bots)
+    except UnicodeDecodeError:
+        return any(bot in user_agent.encode('ascii', 'ignore') for bot in bots)
```

Comment:
Fallback should be removed, either `user_agent` is actually `bytes` and should always be decoded, or `user_agent` is always `str` and the error will never occur. This is likely a remnant of Python 2 code.